### PR TITLE
Feature/Server/Client/Change password

### PR DESCRIPTION
**Pull request by hyejin-seok**
## Describe your changes

- Create change password, Server side
- Create change password, Client side
- Add change password and email validation for edit account in e2e test

## Issue
#78 
#79 

## URL
http://localhost:3000/account/change-password

## Screenshot (if you made the UI)
Screenshot 2024-04-10 at 03 24 24
Screenshot 2024-04-10 at 03 25 15

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings and no errors

### CODE REVIEW

File: client/app/account/components/AccountInfo.tsx [hunk 1]
```diff
@@ -35,16 +35,15 @@ const AccountInfo = () => {
...
         </RouterLink>
       )}
-      {/* TODO: Add functionality for changing password
       <div>
         <h2 className='font-semibold'>Password</h2>
         <RouterLink
-          href='/edit-password'
-          className='text-lg font-normal underline'
+          href='/account/change-password'
+          className='underline underline-offset-4'
```
Comment (kanatagu) [inline client/app/account/components/AccountInfo.tsx:42]: Can you add hover style too? 🙏  Maybe opacity 70 or 80 is good!

File: client/app/account/hooks/useChangePassword.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      confirmNewPassword: '',
+    },
+  })
+
+  const { mutate, isPending } = useChangePasswordMutation()
+
+  const onUpdateAccount = (data: PasswordSchema) => {
+    mutate(data, {
+      onSuccess: () => {
+        queryClient.invalidateQueries({ queryKey: ['me'] })
```
Comment (kanatagu) [inline client/app/account/hooks/useChangePassword.ts:28]: It's better to say `onChansePassword` for the function name. And you can delete the `queryClient.invalidateQueries({ queryKey: ['me'] })` as we did not store in the cache!

File: server/src/user/user.service.ts [hunk 3]
```diff
@@ -136,6 +139,51 @@ export class UserService {
...
+        throw new NotFoundException(`User with ID ${userId} not found.`);
+      }
+
+      const passwordMatch = await compare(
+        dto.currentPassword,
+        user.hashedPassword,
+      );
+
+      if (!passwordMatch) {
+        throw new UnauthorizedException('Invalid credentials');
```
Comment (kanatagu) [inline server/src/user/user.service.ts:160]: I think in this case, it's better to send `The current password is not correct` as the error message!

File: server/src/user/user.service.ts [hunk 4]
```diff
@@ -136,6 +139,51 @@ export class UserService {
...
+      }
+
+      if (dto.newPassword !== dto.confirmNewPassword) {
+        throw new BadRequestException(
+          'New password and confirmation password do not match',
+        );
+      }
```
Comment (kanatagu) [inline server/src/user/user.service.ts:167]: Oh, actually, you don't need to check this. I think if you check if the newPassword and confirmNewPassword match on the client side, you don't need to pass it to the server!

File: server/src/user/dto/change-password-dto.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  @IsNotEmpty()
+  @MinLength(8)
+  @MaxLength(50)
+  newPassword: string;
+
+  @IsString()
+  @IsNotEmpty()
+  @MinLength(8)
+  @MaxLength(50)
+  confirmNewPassword: string;
```
Comment (kanatagu) [inline server/src/user/dto/change-password-dto.ts:20]: You don't need this!

File: client/app/account/api/changePasswordApi.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import apiClient from '@/_api/base'
+import { UserType } from '@/_types'
+import { PasswordSchema } from '../schema'
+
+export const changePasswordApi = async (
+  data: PasswordSchema,
+): Promise<UserType> => {
+  const response = await apiClient.put<UserType>('/me/change-password', data)
```
Comment (kanatagu) [inline client/app/account/api/changePasswordApi.ts:8]: Return type is `{ message: string }`  if you return like this `return { message: 'Password updated successfully.' };`

File: client/app/account/components/AccountInfo.tsx (hunk 1 above)
Comment (hyejin-seok) [reply client/app/account/components/AccountInfo.tsx]: Right! Added it. 
fa7a2caeced095df603d2e0ae9982658e5c753f0

File: client/app/account/hooks/useChangePassword.ts (hunk 2 above)
Comment (hyejin-seok) [reply client/app/account/hooks/useChangePassword.ts]: Fixed it.
8d19e953ba556e4fe9875e45cc47c29408c60f3d

File: server/src/user/user.service.ts (hunk 3 above)
Comment (hyejin-seok) [reply server/src/user/user.service.ts]: Fixed it.
d3c093a061236221f5429c0eb4b6d5312e98515a

File: server/src/user/user.service.ts (hunk 4 above)
Comment (hyejin-seok) [reply server/src/user/user.service.ts]: Oh I see!
716cdb3402c6a0a80e41db82e93ae2a64a3ba0f3

File: server/src/user/dto/change-password-dto.ts (hunk 5 above)
Comment (hyejin-seok) [reply server/src/user/dto/change-password-dto.ts]: Fixed it.
716cdb3402c6a0a80e41db82e93ae2a64a3ba0f3

File: client/app/account/api/changePasswordApi.ts (hunk 6 above)
Comment (hyejin-seok) [reply client/app/account/api/changePasswordApi.ts]: Got it! 
c852d0de6de152c6b0300c576333c89bd95dc47b

**Comment by hyejin-seok** [comment]
I found out that when a user doesn't yet have a house, the user can't change the password. I fixed it. 
c3c5c7d1686783b52e314c7bc752f903d582283d

File: client/app/account/hooks/useChangePassword.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+      currentPassword: '',
+      newPassword: '',
+      confirmNewPassword: '',
+    },
+  })
+
+  const { mutate, isPending } = useChangePasswordMutation()
+
+  const onChansePassword = (data: PasswordSchema) => {
+    mutate(data, {
```
Comment (kanatagu) [inline client/app/account/hooks/useChangePassword.ts:24]: You don't need to pass `confirmNewPassword`. You can just pass except it!

File: client/app/account/api/changePasswordApi.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import apiClient from '@/_api/base'
+import { PasswordSchema } from '../schema'
+
+export const changePasswordApi = async (
+  data: PasswordSchema,
```
Comment (kanatagu) [inline client/app/account/api/changePasswordApi.ts:5]: This data type should be like this
`Omit<PasswordSchemam, 'confirmNewPassword'>` same as the argument of mutationFn in useChangePasswordMutation

File: client/app/account/hooks/useChangePassword.ts (hunk 7 above)
Comment (hyejin-seok) [reply client/app/account/hooks/useChangePassword.ts]: Got it! Thanks!
d2a454df2daf2d0a1857e64502c18f254a7a3d6d

File: client/app/account/api/changePasswordApi.ts (hunk 8 above)
Comment (hyejin-seok) [reply client/app/account/api/changePasswordApi.ts]: Fixed it.
d2a454df2daf2d0a1857e64502c18f254a7a3d6d

**Review by kanatagu — APPROVED** [review]
LGTM!

**Pull request merged**